partner/WHMCS/modules/servers/jetpack/jetpack.php

```diff
@@ -0,0 +1,383 @@
+    } elseif (empty($response)) {
+        throw new Exception('Empty response');
+    }
+
+
+    $decoded_response = json_decode($response);
+    $decoded_response->http_status = $http_code;
+    curl_close($curl);
+
+    return $decoded_response;
+}
+
+/**
+ * Save the next_url for Jetpack activation/setup to the order for the client. If a next url is not set or is empty
+ * and the pending param is true save a message letting the user know a plan is pending and will
+ * be available once the domain the plan is being provisioned for resolves
+ *
+ * @param string $url The next url for activating the Jetpack plan
+ * @param array $params WHMCS params
+ * @param bool $pending If the plan is pending domain resolution.
+ */
+function save_provisioning_details($url, $params, $pending = false)
+{
+    $jetpack_next_url_field = Capsule::table('tblcustomfields')
+        ->where(['fieldname' => 'jetpack_provisioning_details', 'type' => 'product'])->first();
+
+    $details = '';
+    if ($url) {
+        $details = 'URL to Activate Jetpack: ' . $url;
+    } elseif ($pending) {
+        $details = 'The domain did not appear to resolve when provisioning was attempted however a Jetpack plan is 
+        waiting for ' . $params['customfields']['Site URL'] . '. Once DNS resolves please connect the site via 
+        the Jetpack Banner in the sites dashboard';
+    }
+    Capsule::table('tblcustomfieldsvalues')->where(['fieldid' => $jetpack_next_url_field->id])->update([
+        'relid' => $params['model']['orderId'], 'value' => $details]);
+}
+
+/**
+ * Validate that the module was correctly set up when the product was
+ * created by the WHMCS user and that the required Fields/Options for
+ * being able to provision a Jetpack plan are present. Fields validated are
+ *  - Allowed Plans from Plan Custom Field
+ *  - Required Custom Fields
+ *  - Required Config Options
+ *
+ * @param array $params WHMCS params
+ * @return bool|string An error describing what was not correctly included in the setup of the module or True if
+ * everything checks out and provisioning can be attempted.
+ */
+function validate_required_fields(array $params)
+{
+    $allowed_plans = ['free', 'personal', 'premium', 'professional'];
+    $required_custom_fields = ['Plan', 'Site URL', 'Local User', 'jetpack_provisioning_details'];
+
+    foreach ($required_custom_fields as $field) {
+        if (!isset($params['customfields'][$field])) {
+            return 'JETPACK MODULE: The module does not appear to be setup correctly. The required custom field '
+                . $field . ' was not setup when the product was created.
+				Please see the module documentation for more information';
+        }
+    }
+
+    if (!in_array(strtolower($params['customfields']['Plan']), $allowed_plans)) {
+        return 'JETPACK MODULE: The module does not appear to be setup correctly. ' .
+            $params['customfields']['Plan'] . ' is not an allowed plan';
```

Comment:
This value is used in the WHCMS special functions, do we know whether they are used in HTML context and do we need to escape the user-supplied values here?

Reply:
@olska my question is if `$params['customfields']['Plan']` is `<script>alert('hi!');</script>`, would it be executed on a HTML page somewhere?

Reply:
@nb both the user and host supplied custom field values are escaped already when saved so this alert will not be executed on any HTML pages anywhere. We do not need to escape these in the context of the module which currently uses core WHMCS functions however if anything custom is done revisiting this would be necessary.

Reply:
@nb I also did verify this in the database table for custom fields to be sure things are as expected.

Reply:
@olska maybe I am missing something, I am sorry for the repeated questions, but where exactly is the value that a user enters in the Plan field (even if it's a drop-down, it's trivial for the user to send any value they want) escaped before saving? Is it somewhere inside WHMCS? Or somewhere in our code that I have missed?

Reply:
@nb no worries about the questions and thanks for asking for clarification. It's a legitimate concern so I do want to make sure we are on the same understanding and if there's a change that needs to be made to resolve a concern it is addressed. The custom fields are not created or managed by the module. The host sets these up in the WHMCS product settings as part of the set up steps. A user can send any value they want but these are managed and saved by WHMCS and escaped accordingly. WHMCS code is specifically handling this right now and the values are escaped from wherever they are created including on the order page when the user sends their details to the host to get a Jetpack plan.